### ISSUE TITLE
Possible to add more (pluggable) QLs?

### ISSUE THREAD

**Issue by bionicles** [terminusdb/terminusdb#281]
**Is your feature request related to a problem? Please describe.**
This looks really interesting but I would want to be able to use legit graph query language such as openCypher, gremlin, datalog

**Describe the solution you'd like**
(terminus:DB)-[:supports]->(gql:Query language)

Note the SQL standard is evolving to eventually include GQL (that is NOT GraphQL) so if Terminus began to support a reasonably close dialect of openCypher then it could eventually become the only permissive-license open-source graph database to support this new paradigm. All the others are closed source or lack permissive license which means they couldn't really be embedded in a product

On the other hand, Gremlin is easier to implement, so that could be cranked out faster

**Describe alternatives you've considered**
RedisLabs Cloud supports RedisGraph now, even on the free tier. That's super compelling but I'm keen to embed a graph database in my project. I also like Rust a lot

**Additional context**
Oracle PGQL is perhaps the best as it has schemas to ensure data integrity https://pgql-lang.org/

The Linked Data Benchmark Council has standard tests for Graph DBs
http://ldbcouncil.org/

There's a beta cypher parser written in C with a permissive license which could be rewritten in rust https://cleishm.github.io/libcypher-parser/ and would benefit the graph DB community a great deal

GQL standard has a website https://www.gqlstandards.org/
and published scope/features document
https://s3.amazonaws.com/artifacts.opencypher.org/website/materials/sql-pg-2018-0046r3-GQL-Scope-and-Features.pdf

openCypher grammar, integration tests are available here https://github.com/opencypher/openCypher

I'm unclear how database query engines work but I'll look into it

**Comment by luke-feeney** [comment]
Thanks @bionicles - query languages are always a bit fraught. WOQL is a datalog, so it is similar to graph query languages used by other upstart graph databases like Grakn, Datomic, and OpenCrux. WOQL is based on prolog so its a proper datalog. We think this is the way of the future (with python SDKs doing some of the work), especially for complex data. 

We debated using SPARQL initially; however, we found SPARQL to have a number of shortcomings that we wanted to see addressed. We wanted tighter integration with OWL and schema-awareness, so if we were going to need to fundamentally alter the semantics, it didn’t seem that important to start with SPARQL. We wanted better composability and SPARQL feels quite ad-hoc.

We also assessed Cypher and found that it lacked some of the features we needed. It is also primarily geared towards property graphs and not RDF graphs. It also does not enable the smooth SDK manipulation that WOQL facilitates by using JSON-LD as a query interchange format.

We think that the Python and JavaScript SDKs for TerminusDB are the main way that users will interact with the database, so we are focused on getting them right rather than adding more QLs for the moment.

I fear that the GQL standards process will drag on forever and could end up with something that nobody really wants. Neo4j seemed to be driving it for a while, but are less prominent lately. 

If we were adding another QL, I'd favor a SQL wrapper of some sort. Better to go towards the biggest possible pool (as graph query is still pretty niche).

**Comment by bionicles** [comment]
What about just letting us use Prolog / Datalog instead?  They're still hard to read but at least supported more widely, and it looks like Terminus uses a lot of Prolog anyway...

WOQL looks verbose and hard to follow for me, in comparison with cypher which is elegant and easy to read. Also, WOQL only works with TerminusDB, so it's a lot harder to communicate with stakeholders because I'd be using this one-off language. 

The main reason I'd use a graph database is to use Cypher over SQL, but I wanted to avoid Neo4j because Java sucks, and I like RedisGraph, but it doesn't support temporal queries, and a weird license.  Also, redisgraph doesn't yet support JSON. ArangoDB is the only db which supports graphs and json, but they don't support temporal queries and have their own one-off language AQL, and pretty scattered docs.  I'd love to use a temporal graph database which uses Cypher and supports static type declarative schema derived from Rust structs, as well as JSON, which is everywhere and pretty required to deal with.  Datomic might be nice but alas, more Java.  Tigergraph is way too expensive and proprietary. Oracle Graph DB looks cool, since you can use a schema, but ugh, oracle, of course it's more java.  Memgraph, not great cloud offerings, more "fully managed" (but not managed autoscaling?). Neptune, proprietary, also "fully managed" (but not managed autoscaling?) and gremlin.  Janusgraph, java, gremlin.  CosmosDB, gremlin, microsoft, no thanks.  Catching my drift here?  There's no truly great open-source permissive-license strongly-typed temporal openCypher graph database. MAYBE Apache AGE, but it's quite new, I'd be scared.  

I accept the fact cypher is difficult to implement since it requires a parser and assorted components.  To me, this is worth the effort, because many graph DBs now support it, and it is a simple and easy language to read.  SQL is obviously a nice feature but defeats the purpose of going toward a graph database.  Heck, even Gremlin would be better than WOQL.  Just one man's opinion: I love the inner workings and capabilities of TerminusDB. Dislike how I'd query it because WOQL doesn't make semantic sense. 


Also, I'm using Rust, is there a client for the crab language?

Anyway... It's hard to pick a database. There are many options and none really stand out to me. I'll just suck it up and use GraphQL on MySQL, at least then I can generate DDL from Rust structs and do declarative migrations with Skeema.

**Comment by GavinMendelGleason** [comment]
It is possible technically to add other query languages but we have no plans to do so presently. 

It is possible to query using WOQL embedded in prolog - we use it internally, but it would be nice to make this a more external interface. There are big advantages to having queries be an AST however, rather than strings as is the case in many other query languages, and this can enable tighter integration in your language of choice. 

I'm not sure that "WOQL doesn't make much semantic sense" is reasonable given that it does have a clear semantics based on datalog.

**Comment by sandias42** [comment]
Just felt like chiming in my emphatic agreement with the OP.  I was very excited to try terminusdb at my company after searching for quite a long time for an elegant schema-enforcing graph db. I read through all the technical internals blogs, loved them, and then encountered the example WOQL query from that blog post

```WOQL.("v:Living Person Record ID", "status", "alive")
   	    .path("v:Living Person Record ID", "parent+", "v:Italian Ancestor", "v:Ancestry Line")
   	    .triple("v:Italian Ancestor", "date_born", "v:Date of Birth")
   	    .less("v:Date of Birth", 1850)
   	    .triple("v:Italian Ancestor", "country_born", WOQL.string("Italy"))
   	    .read_document("v:Living Person Record ID", "v:Full Record")```
   	    
  ...and completely deflated. The *semantics* of WOQL (as described in the blog post) make sense to me and strike me as elegant, but this syntax is a dealbreaker for me. I'm excited enough about terminus despite this that I'm in the process of searching through supported alternatives to querying and even looking for random github projects with alternative QLs for terminus.
  
 I find the syntax of cypher extremely intuitive and fun to read and write. If it (or something aesthetically similar) were supported I'd move my company to terminus over our current graph db.
